xcode/Subconscious/Shared/UIKit/BlockEditor/BlockEditorViewController.swift

```diff
@@ -0,0 +1,1211 @@
+                ),
+                collectionView.bottomAnchor.constraint(
+                    equalTo: guide.bottomAnchor
+                ),
+            ])
+            
+            self.store.connect(self)
+        }
+        
+        @objc private func onLongPress(_ gesture: UIGestureRecognizer) {
+            switch gesture.state {
+            case .ended:
+                let point = gesture.location(in: self.collectionView)
+                store.send(.longPress(point))
+            default:
+                break
+            }
+        }
+
+        @objc private func onTap(_ gesture: UIGestureRecognizer) {
+            switch gesture.state {
+            case .ended:
+                let point = gesture.location(in: self.collectionView)
+                store.send(.tap(point))
+            default:
+                break
+            }
+        }
+
+        /// Return 2 sections
+        /// - 1 blocks
+        /// - 2 "footer" containing related notes
+        func numberOfSections(in collectionView: UICollectionView) -> Int {
+            Section.allCases.count
+        }
+
+        /// Provides a count for data in a particular section
+        func collectionView(
+            _ collectionView: UICollectionView,
+            numberOfItemsInSection section: Int
+        ) -> Int {
+            switch Section(rawValue: section) {
+            case .blocks:
+                return store.state.blocks.count
+            case .appendix:
+                return 1
+            default:
+                return 0
+            }
+        }
+        
+        /// Provides a UICollectionViewCell for an index path.
+        /// This method is responsible for looking up data from a data source
+        /// (state in our case) using the index path.
+        func collectionView(
+            _ collectionView: UICollectionView,
+            cellForItemAt indexPath: IndexPath
+        ) -> UICollectionViewCell {
+            switch Section(rawValue: indexPath.section) {
+            case .blocks:
+                return blockCell(collectionView, forItemAt: indexPath)
+            case .appendix:
+                return appendixCell(collectionView, forItemAt: indexPath)
+            default:
+                return errorCell(collectionView, forItemAt: indexPath)
+            }
+        }
+        
+        /// Error cells are dequeued when we don't know what else to display.
+        ///
+        /// Error cells should never be displayed in practice, but we must
+        /// have something to dequeue since `collectionView(_:cellForItemAt:)`
+        /// requires a cell to be returned for every case.
+        private func errorCell(
+            _ collectionView: UICollectionView,
+            forItemAt indexPath: IndexPath
+        ) -> UICollectionViewCell {
+            let cell = collectionView.dequeueReusableCell(
+                withReuseIdentifier: ErrorCell.identifier,
+                for: indexPath
+            ) as! ErrorCell
```

Comment:
Can/should this be annotated with `throws` since it contains `as!`?

Reply:
This is an unfortunate artifact of a UIKit API designed for Objective C that is unergonomic in Swift.

- Cells are meant to be dequeued from a shared pool of reusable cells (managed by the collection view not us).
- Reusable cells are registered via a key, and there is no way for the type system to guarantee that you actually registered the cell type.
- Meanwhile, it is required that `func collectionView(_ collectionView: UICollectionView,cellForItemAt indexPath: IndexPath) -> UICollectionViewCell` return a UICollectionViewCell. It is non-optional.

So we are reduced to force-unwrapping the cell after dequeuing. As long as the cell type is registered, this invariant will be upheld, and luckily if it is not registered, you get an instant crash, so you know it.